placement/algo/algo.go

```diff
@@ -0,0 +1,482 @@
+	var err error
+	if ph, leavingHostShards, err = newRemoveHostPlacementHelper(ps, leavingHost); err != nil {
+		return nil, err
+	}
+	addingHostShards := newEmptyHostShardsFromHost(addingHost)
+	var shardsUnassigned []uint32
+	// move shards from leaving host to adding host
+	for _, shard := range leavingHostShards.Shards() {
+		if ph.canAssignHost(shard, leavingHostShards, addingHostShards) {
+			ph.removeShardFromHost(shard, leavingHostShards)
+			addingHostShards.addShard(shard)
+		} else {
+			shardsUnassigned = append(shardsUnassigned, shard)
+		}
+	}
+	// if there are shards that can not be moved to adding host
+	// distribute them to the cluster
+	if err := ph.placeShards(shardsUnassigned, leavingHostShards); err != nil {
+		return nil, err
+	}
+
+	// add the adding host to the cluster and bring its load up to target load
+	cl := ph.generatePlacement()
+	return a.addHostShards(cl, addingHostShards)
+}
+
+func isHostAbsent(ps placement.Snapshot, h placement.Host) bool {
+	for _, hs := range ps.HostShards() {
+		if hs.Host().Address == h.Address {
+			return false
+		}
+	}
+	return true
+}
+
+// shardAwareDeploymentPlanner plans the deployment so that as many hosts can be deployed
+// at the same time without making more than 1 replica of any shard unavailable
+type shardAwareDeploymentPlanner struct {
+}
+
+func newShardAwareDeploymentPlanner() placement.DeploymentPlanner {
+	return shardAwareDeploymentPlanner{}
+}
+
+func (dp shardAwareDeploymentPlanner) DeploymentSteps(ps placement.Snapshot) [][]placement.HostShards {
+	ph := newReplicaPlacementHelper(ps, ps.Replicas())
+	var steps [][]placement.HostShards
+	for ph.hostHeap.Len() > 0 {
+		h := heap.Pop(ph.hostHeap).(*hostShards)
+		var parallel []placement.HostShards
+		parallel = append(parallel, h)
+		var tried []*hostShards
+		for ph.hostHeap.Len() > 0 {
+			tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+			if !h.isSharingShard(*tryHost) {
+				parallel = append(parallel, tryHost)
+			} else {
+				tried = append(tried, tryHost)
+			}
+		}
+		for _, host := range tried {
+			heap.Push(ph.hostHeap, host)
+		}
+		steps = append(steps, parallel)
+	}
+	return steps
+}
+
+type placementHelper struct {
+	hostHeap       *hostHeap
+	shardToHostMap map[uint32]map[*hostShards]struct{}
+	rackToHostsMap map[string]map[*hostShards]struct{}
+	rf             int
+	uniqueShards   []uint32
+	hostShards     []*hostShards
+}
+
+func newInitPlacementHelper(hosts []placement.Host, ids []uint32) *placementHelper {
+	emptyPlacement := newEmptyPlacement(hosts, ids)
+	return newPlaceShardingHelper(emptyPlacement, emptyPlacement.rf+1, true)
+}
+
+func newReplicaPlacementHelper(s placement.Snapshot, targetRF int) *placementHelper {
+	ps := newPlacementFromGenericSnapshot(s)
+	return newPlaceShardingHelper(ps, targetRF, true)
+}
+
+func newAddHostShardsPlacementHelper(s placement.Snapshot, hs *hostShards) (*placementHelper, *hostShards, error) {
+	var hss []*hostShards
+	for _, phs := range s.HostShards() {
```

Comment:
Yeah so if we have a `HasHost` method on `Snapshot` interface this could be simplified